src/NuGet.Core/NuGet.Packaging/Signing/Archive/ReadOnlyBufferedStream.cs

```diff
@@ -0,0 +1,365 @@
+            get
+            {
+                ThrowIfDisposed();
+
+                return _stream.ReadTimeout;
+            }
+            set
+            {
+                ThrowIfDisposed();
+
+                _stream.ReadTimeout = value;
+            }
+        }
+
+        public override int WriteTimeout
+        {
+            get
+            {
+                ThrowIfDisposed();
+
+                return _stream.WriteTimeout;
+            }
+            set
+            {
+                ThrowIfDisposed();
+
+                _stream.WriteTimeout = value;
+            }
+        }
+
+        public ReadOnlyBufferedStream(Stream stream, bool leaveOpen) :
+            this(stream, leaveOpen, _defaultBufferSize)
+        {
+        }
+
+        public ReadOnlyBufferedStream(Stream stream, bool leaveOpen, int bufferSize)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (!stream.CanRead)
+            {
+                throw new ArgumentException(Strings.StreamMustBeReadable, nameof(stream));
+            }
+
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException(Strings.StreamMustBeSeekable, nameof(stream));
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize));
+            }
+
+            _buffer = new byte[bufferSize];
+            _stream = stream;
+            _leaveOpen = leaveOpen;
+
+            // Repeated calls to the Length property were found to be costly.
+            _length = new Lazy<long>(() => _stream.Length);
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            return base.CopyToAsync(destination, bufferSize, cancellationToken);
+        }
+
+        public override void Flush()
+        {
+            ThrowIfDisposed();
+
+            // This is a read-only stream.
+            throw new NotSupportedException();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            // This is a read-only stream.
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ThrowIfDisposed();
+
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (count - offset > buffer.Length ||
+                count + offset > buffer.Length)
```

Comment:
why count + offset?

Reply:
Even I dont get this condition `count - offset > buffer.Length`

Reply:
I also don´t get why `count - offset > buffer.Length`, and also given that count and offset are always positive values `count - offset > buffer.Length` can never be false if `count + offset > buffer.Length` is true, therefore having both checks is redundant.

Reply:
I've revised this to:

    if (count > buffer.Length - offset)